src/plugins/Trader/UI/PriceChart.tsx

```diff
@@ -0,0 +1,137 @@
+import React, { useRef, useEffect } from 'react'
+import * as d3 from 'd3'
+import type { Stat } from '../types'
+import { makeStyles, Theme, createStyles, CircularProgress, Typography } from '@material-ui/core'
+
+const DEFAULT_WIDTH = 460
+const DEFAULT_HEIGHT = 250
+const DEFAULT_MARGIN = {
+    top: 32,
+    right: 16,
+    bottom: 32,
+    left: 16,
+}
+
+const useStyles = makeStyles((theme: Theme) => {
+    return createStyles({
+        root: {
+            position: 'relative',
+        },
+        svg: {},
+        progress: {
+            bottom: theme.spacing(1),
+            right: theme.spacing(1),
+            position: 'absolute',
+        },
+        placeholder: {
+            paddingTop: theme.spacing(14),
+        },
+    })
+})
+
+export interface PriceChartProps {
+    stats: Stat[]
+    loading?: boolean
+    width?: number
+    height?: number
+    children?: React.ReactNode
+}
+
+export function PriceChart(props: PriceChartProps) {
+    const classes = useStyles()
+    const svgRef = useRef<SVGSVGElement>(null)
+
+    // define dimensions
+    const {
+        width = DEFAULT_WIDTH - DEFAULT_MARGIN.left - DEFAULT_MARGIN.right,
+        height = DEFAULT_HEIGHT - DEFAULT_MARGIN.top - DEFAULT_MARGIN.bottom,
+    } = props
+    const canvasWidth = width + DEFAULT_MARGIN.left + DEFAULT_MARGIN.right
+    const canvasHeight = height + DEFAULT_MARGIN.top + DEFAULT_MARGIN.bottom
+
+    // process data
+    const data = props.stats.map(([date, price]) => ({
+        date: new Date(date),
+        value: price,
+    }))
+
+    useEffect(() => {
+        if (!svgRef.current) return
+
+        // empty the svg
+        svgRef.current.innerHTML = ''
+
+        // render svg if necessary
+        if (!props.stats.length) return
+
+        // contine to create the chart
+        const svg = d3
```

Comment:
Ah... I've heard that code using d3 is hard to maintain. Is there are a more declartive library? Or you said you want to use their library to draw, does it failed so you have to draw the chart yourself?

Reply:
It's easy to use and extend. I will extract those codes follows one graph for one file.